### PR TITLE
Adds AI Integrity restorer to program datums, various modular computer improvements

### PR DESCRIPTION
**Pull request by Shadowlight213**
:cl:
add: Added the AI integrity restorer as a modular computer program
add: Added an AI intelliCard slot. Insert an Intellicard into it to be able to use the restoration program.
fix: Fixes Alarm program detecting ruins
fix: Fixes being unable to toggle the card reader module power
tweak: The downloader will now tell you if a program is incompatible with your hardware
/:cl:

Slight improvement to computer part destroy hopefully. I couldn't fully test it due to the reference detector being broken due to a possible byond bug.  http://www.byond.com/forum/?post=2172860
Also fixed the intelliCard tgui braindead check.

### CODE REVIEW

**Comment by GunHog** [comment]
Does the RD's console or tablet come with this module and program installed? It will significantly increase the chances of this program being used if the RD already has it on his console or tablet roundstart. Else, most will just use the current AI fixer. (Which even comes with sounds, now! This is almost a downgrade due to that!)

**Comment by Shadowlight213** [comment]
@GunHog It's in the console in the rd's office, roundstart, yes.
Also, I restricted the program to only be runnable on consoles.

**Comment by ToadStar** [comment]
Would it be too much to ask for more functional programs from baycode to be ported? Stuff like the crew monitor, med/sec records, the _incredible_ camera console program (Seriously, look at it sometime, it's fucking amazing compared to the giantass list we have now), because honestly, the modular computers don't really do much that can't already be done with the current circuitboard based computers.

**Comment by Shadowlight213** [comment]
@ToadStar I've been thinking about that. The thing is, I'd need to figure out how to port their map wrapper to tgui, as well as the sec camera map thing itself.
Really, I'd love to port everything, and for the most part it's not really hard.
It's just <i>really</i> tedious. It took me almost a whole day of nonstop coding just to do this one program. When I did the initial port, I was able to convert a lot of stuff pretty much directly.
I'll try to get around to adding more stuff.

**Comment by ExcessiveUseOfCobblestone** [comment]
who removed the id slot / id changer app for the hop, literally the best thing about the tablets.

**Comment by Shadowlight213** [comment]
@ExcessiveUseOfCobblestone they never had it? I think the app itself is limited to at least laptops.
I'm actually not sure why tablets have the id card slot.

**Comment by GunHog** [comment]
I often like to ask the AI to download the ID Program to my tablet when I play RD; I can then use it as a portable Science restricted ID console (like the one in my office). It does NOT allow the granting of all-access unless a HoP/Captain ID inserted. Every head should have one, IMO.

**Comment by Shadowlight213** [comment]
@ExcessiveUseOfCobblestone he does have it though?

**Comment by ExcessiveUseOfCobblestone** [comment]
huh last time I tried it the ID changer app wasn't there.

**Comment by Shadowlight213** [comment]
@ExcessiveUseOfCobby it doesn't come pre-installed. Did you check the downloader?

**Comment by ExcessiveUseOfCobblestone** [comment]
yeah, maybe I was just misreading it like a dummy.

**Comment by GunHog** [comment]
All heads should have it on their tablets imo, for the non HoP heads, they funciton just like the department consoles in each head office!

File: code/modules/modular_computers/file_system/programs/airestorer.dm [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	if(computer)
+		ai_slot = computer.all_components[MC_AI]
+
+	if(computer && ai_slot && ai_slot.check_functionality() && ai_slot.enabled && ai_slot.stored_card && ai_slot.stored_card.AI)
+		return ai_slot.stored_card.AI
+	return null
+
+/datum/computer_file/program/aidiag/ui_act(action, params)
+	if(..())
+		return 1
```
Comment (optimumtact) [inline code/modules/modular_computers/file_system/programs/airestorer.dm]: TRUE/FALSE defines?

**Review by optimumtact — CHANGES_REQUESTED** [review]
Some minor concerns

File: code/modules/modular_computers/computers/item/computer.dm [hunk 2]
```diff
@@ -94,6 +99,19 @@
...
+/obj/item/device/modular_computer/proc/eject_card()
+	set name = "Eject Intellicard"
+	set category = "Object"
+	set src in view(1)
+
+	if(issilicon(usr))
+		return
+	var/obj/item/weapon/computer_hardware/ai_slot/ai_slot = all_components[MC_AI]
+	if(usr.canUseTopic(src))
+		ai_slot.try_eject(, usr,1)
```
Comment (Shadowlight213) [inline code/modules/modular_computers/computers/item/computer.dm]: I'm calling the try eject proc?
When whoever it was rewrote hardware, they had the id slot work this way.

File: code/modules/modular_computers/computers/item/computer.dm [hunk 3]
```diff
@@ -94,6 +99,19 @@
 		card_slot.try_eject(, usr)
 
 // Eject ID card from computer, if it has ID slot with card inside.
+/obj/item/device/modular_computer/proc/eject_card()
+	set name = "Eject Intellicard"
+	set category = "Object"
+	set src in view(1)
```
Comment (Shadowlight213) [inline code/modules/modular_computers/computers/item/computer.dm]: ? I don't understand.
I'm basing this off of the proc for the id card which was done by @Razharas a maintainer.
Why would it be wrong?

File: code/modules/modular_computers/computers/item/computer.dm (hunk 2 above)
Comment (optimumtact) [inline code/modules/modular_computers/computers/item/computer.dm]: why is there a stray comma there?

File: code/modules/modular_computers/computers/item/computer.dm (hunk 3 above)
Comment (optimumtact) [inline code/modules/modular_computers/computers/item/computer.dm]: it's not even needed, this is copypasta from old verb code

File: code/modules/modular_computers/computers/item/computer.dm (hunk 2 above)
Comment (Shadowlight213) [inline code/modules/modular_computers/computers/item/computer.dm]: It's not stray. The first argument is null, same way that the id card one handles it.

File: code/modules/modular_computers/computers/item/computer.dm (hunk 3 above)
Comment (Shadowlight213) [inline code/modules/modular_computers/computers/item/computer.dm]: K

File: code/modules/modular_computers/computers/item/computer.dm (hunk 2 above)
Comment (optimumtact) [inline code/modules/modular_computers/computers/item/computer.dm]: can you put null then, because it looks like a mistype

**Comment by Shadowlight213** [comment]
fixed

**Review by optimumtact — APPROVED** [review]

**Comment by KorPhaeron** [comment]
Having a mobile cloner in your backpack for the AI sounds kind of off

Machines/computers are pretty powerful, and it used to be that these powers were tied to a location (or your ability to rob that location and deconstruct/reconstruct it elsewhere).

The trend of things like the HoP being able to stuff his office in his bag feels like pretty bad power/storage creep

**Comment by GunHog** [comment]
> Having a mobile cloner in your backpack for the AI sounds kind of off

The AI slot is only compatible with the stationary console, according to a previous post.

>The trend of things like the HoP being able to stuff his office in his bag feels like pretty bad power/storage creep

If modular computers do not have useful programs, nobody is going to want one! Besides, being HoP and never leaving the office is unrewarding and boring. That also opens him to the chance of making really poor decisions, such as changing an ID in the middle of a hallway..one slip and your fancy tablet is gone!

**Pull request merged**